### PR TITLE
Log-scale layout of heatmap bricks for log-scale axes (fixes #5991)

### LINKED ISSUE #5991: heatmap bricks on log axes
From the discussion that came up in #5985 - that PR proposed 4 contour plots - on a mixture of log and linear axes - that should look identical but didn't, until fixed in the PR. If we make the same plots as a smoothed heatmap, they look identical as far as the smoothing goes, but we lose the first row of bricks on each log axis because its far edge would be at a negative coordinate:

> Also closer to what we already do with smoothed heatmaps (the same mock after `Plotly.restyle(gd,{'type':'heatmap', 'zsmooth':'best'})` - the four variants look slightly different but I think this is actually a symptom of some funny and potentially incorrect things we're doing with extending beyond the data values to make the outer edges of the heatmap bricks - we may want to adapt this for log axes too, especially as this results in us dropping the first row/column in this case!):
> Screen Shot 2021-10-18 at 12 06 13 PM

Can we do a better job of picking brick edges in this case?

The figure above would be:

```
{
  "data": [
    {
      "x": [1,2,3,4,5],
      "y": [1,2,3,4,5],
      "z": [
        [0 ,   3.1250,   12.5000,   28.1250,   50.0000],
        [3.1250,    6.2500,   15.6250,   31.2500,   53.1250],
        [12.5000,   15.6250,   25.0000,   40.6250,   62.5000],
        [28.1250,   31.2500,   40.6250,   56.2500,   78.1250],
        [50.0000,   53.1250,   62.5000,   78.1250,  100.0000]
      ],
      "zsmooth": "best",
      "type": "heatmap"
    },
    {
      "x": [1,10,100,1000,10000],
      "y": [1,2,3,4,5],
      "z": [
        [0 ,   3.1250,   12.5000,   28.1250,   50.0000],
        [3.1250,    6.2500,   15.6250,   31.2500,   53.1250],
        [12.5000,   15.6250,   25.0000,   40.6250,   62.5000],
        [28.1250,   31.2500,   40.6250,   56.2500,   78.1250],
        [50.0000,   53.1250,   62.5000,   78.1250,  100.0000]
      ],
      "zsmooth": "best",
      "type": "heatmap",
      "showscale": false,
      "xaxis": "x2"
    },
    {
      "x": [1,2,3,4,5],
      "y": [1,10,100,1000,10000],
      "z": [
        [0 ,   3.1250,   12.5000,   28.1250,   50.0000],
        [3.1250,    6.2500,   15.6250,   31.2500,   53.1250],
        [12.5000,   15.6250,   25.0000,   40.6250,   62.5000],
        [28.1250,   31.2500,   40.6250,   56.2500,   78.1250],
        [50.0000,   53.1250,   62.5000,   78.1250,  100.0000]
      ],
      "zsmooth": "best",
      "type": "heatmap",
      "showscale": false,
      "yaxis": "y2"
    },
    {
      "x": [1,10,100,1000,10000],
      "y": [1,10,100,1000,10000],
      "z": [
        [0 ,   3.1250,   12.5000,   28.1250,   50.0000],
        [3.1250,    6.2500,   15.6250,   31.2500,   53.1250],
        [12.5000,   15.6250,   25.0000,   40.6250,   62.5000],
        [28.1250,   31.2500,   40.6250,   56.2500,   78.1250],
        [50.0000,   53.1250,   62.5000,   78.1250,  100.0000]
      ],
      "zsmooth": "best",
      "type": "heatmap",
      "showscale": false,
      "xaxis": "x2",
      "yaxis": "y2"
    }
  ],
  "layout": {
    "xaxis": {"domain": [0, 0.45]},
    "xaxis2": {"domain": [0.55, 1], "type": "log"},
    "yaxis": {"domain": [0, 0.45]},
    "yaxis2": {"domain": [0.55, 1], "type": "log"},
    "margin": {"l": 50, "b": 50, "r": 110, "t": 10},
    "width": 500,
    "height": 400
  }
}
```

### PR DESCRIPTION
**Pull request by andrew-matteson**
This fixes the heatmap brick layout identified in #5991 . The logic of `make_bounds_array` is extended with logic that replicates the linear scale logic in log space.

This PR fixes two problems in my use of the library:
1. Log-scale heatmaps would not render blocks that calculate a bound < 0.
2. Log-scale heatmaps would not have evenly sized blocks.

I've added a mock and baseline that behave correctly.

### CODE REVIEW

File: test/image/mocks/heatmap-log-scale.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
```
Comment (archmoj) [inline test/image/mocks/heatmap-log-scale.json:1]: Please rename this mock and baseline to start with `zz-` as `test/image/mocks/zz-heatmap-log-scale.json`
That would help avoid a bug in our current parallel image test system.

Comment (andrew-matteson) [reply test/image/mocks/heatmap-log-scale.json]: renamed

**Comment by andrew-matteson** [comment]
@archmoj Can you let me know anything that needs to happen to get this merged?

**Comment by archmoj** [comment]
Thanks for the fix. :1st_place_medal: 
Looking very good to me.
Over to @alexcjohnson

**Comment by alexcjohnson** [comment]
@archmoj after my two comments [above](https://github.com/plotly/plotly.js/pull/6819#discussion_r1458048611) this looks good. Thanks @andrew-matteson!

There's still an issue with `histogram2d` failing on log axes if the bins go past zero, but that's a trickier problem, we need to decide if the bins should also be spaced geometrically (and if so support binning that way) but that feels more like a new feature than a fix like the rest of this PR, since it will alter the data to be displayed, not just the details of how it's displayed - so it doesn't belong here - and is probably a much lower priority anyway, I don't think `histogram2d` gets a whole lot of use.

**Comment by andrew-matteson** [comment]
@alexcjohnson I think my fixes cover your cases. LMK if you have further thoughts.

**Comment by andrew-matteson** [comment]
There is a webgl-jasmine failure that I don't think is related to the recent changes. If it is, could you let me know what's going on? I can't quite deduce from the error messages.

**Comment by andrew-matteson** [comment]
@alexcjohnson I got the tests to rerun and they passed. LMK if this needs anything else.

File: CHANGELOG.md [hunk 2]
```diff
@@ -572,7 +572,7 @@ where X.Y.Z is the semver of most recent plotly.js release.
    with thanks to Kitware and @waxlamp for kicking off this effort.
 
 ### Changed
- - Improve drawing the contour lines in non-linear space e.g. on log axes [[#5985](https://github.com/plotly/plotly.js/pull/5985)], with thanks to @andrew-matteson for the contribution!
+ - Improve drawing the contour lines in non-linear space e.g. on log axes [[#5985](https://github.com/plotly/plotly.js/pull/5985)], with thanks to @andrew-matteson for the contribution!!
```
Comment (archmoj) [inline CHANGELOG.md:575]: Do you want to revert this change?

Comment (andrew-matteson) [reply CHANGELOG.md]: reverted

File: src/traces/heatmap/make_bound_array.js [hunk 3]
```diff
@@ -25,10 +25,20 @@ module.exports = function makeBoundArray(trace, arrayIn, v0In, dvIn, numbricks,
             if(isContour || isGL2D) arrayOut = arrayIn.slice(0, numbricks);
             else if(numbricks === 1) {
                 arrayOut = [arrayIn[0] - 0.5, arrayIn[0] + 0.5];
```
[the remark this answers is not in the document]
Comment (andrew-matteson) [reply src/traces/heatmap/make_bound_array.js]: fixed

**Comment by andrew-matteson** [comment]
@archmoj @alexcjohnson I'd love to get this in for an upcoming release my team has. Is there anything  I need to do to wrap this up?

**Comment by archmoj** [comment]
Thanks very much for the fix. 
:dancer:
We are planning to have a minor plotly.js release next week which would include this.

**Comment by andrew-matteson** [comment]
Thank you! I appreciate this product so much :-)
